internals/workloads/workloads.go

```diff
@@ -120,98 +108,104 @@ func (ws *WorkloadsSection) Validate() error {
-	copied.GroupID = copyPtr(w.GroupID)
-	return &copied
-}
-
-func (w *Workload) Merge(other *Workload) {
-	if len(other.Environment) > 0 {
-		w.Environment = makeMapIfNil(w.Environment)
-		maps.Copy(w.Environment, other.Environment)
-	}
-	if other.UserID != nil {
-		w.UserID = copyPtr(other.UserID)
-	}
-	if other.User != "" {
-		w.User = other.User
-	}
-	if other.GroupID != nil {
-		w.GroupID = copyPtr(other.GroupID)
+func (*WorkloadsSectionExtension) ParseSection(data yaml.Node) (plan.Section, error) {
+	workloads := &WorkloadsSection{}
+	if err := plan.SectionDecode(&data, workloads); err != nil {
+		return nil, &plan.FormatError{
+			Message: fmt.Sprintf(`cannot parse the "workloads" section: %v`, err),
+		}
 	}
-	if other.Group != "" {
-		w.Group = other.Group
+	for name, workload := range workloads.Entries {
+		if workload != nil {
+			workload.Name = name
+		}
 	}
+	return workloads, nil
 }
 
-func (w *Workload) Equal(other *Workload) bool {
-	if !maps.Equal(w.Environment, other.Environment) {
-		return false
-	}
-
-	uid, gid, err := osutil.NormalizeUidGid(w.UserID, w.GroupID, w.User, w.Group)
-	if err != nil {
-		return reflect.DeepEqual(w, other)
+func (ext *WorkloadsSectionExtension) ValidatePlan(p *plan.Plan) error {
+	// The following will panic if the "ws" section is not a WorkloadsSection
+	ws := p.Sections[WorkloadsField].(*WorkloadsSection)
+	for name, service := range p.Services {
+		if service.Workload == "" {
+			continue
+		}
+		if _, ok := ws.Entries[service.Workload]; !ok {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q: not defined for service %q", service.Workload, name),
+			}
+		}
+		if service.UserID != nil || service.User != "" {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("plan service %q cannot have user information and a workload at the same time", name),
+			}
+		}
+		if service.GroupID != nil || service.Group != "" {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("plan service %q cannot have group information and a workload at the same time", name),
+			}
+		}
 	}
-	otherUID, otherGID, err := osutil.NormalizeUidGid(other.UserID, other.GroupID, other.User, other.Group)
-	if err != nil {
-		return reflect.DeepEqual(w, other)
+	for name, workload := range ws.Entries {
+		if _, _, err := osutil.NormalizeUidGid(workload.UserID, workload.GroupID, workload.User, workload.Group); err != nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q: %v", name, err),
+			}
+		}
 	}
-	if uid != nil && gid != nil && otherUID != nil && otherGID != nil {
-		return *uid == *otherUID && *gid == *otherGID
+	if ext.currentPlan != nil {
+		currentWorkloads := ext.currentPlan.Sections[WorkloadsField].(*WorkloadsSection)
+		if !reflect.DeepEqual(currentWorkloads.Entries, ws.Entries) {
+			return &plan.FormatError{
+				Message: "cannot change workloads once the plan has been loaded",
+			}
+		}
 	}
-	return reflect.DeepEqual(w, other)
+	ext.currentPlan = p
```

Comment:
I would not have guessed that the method called "ValidatePlan" also mutates the `ext` object. 

I suggest either 
* changing the name of the function to reflect this, because ultimately it's the most important thing that's happening in this function or 
* separating the implementation of validation and this assignment into a different function.